### PR TITLE
Refactored DateAndTime to remove POCO and Logging Dependencies

### LINKED ISSUE #20523: Remove logging and POCO dependencies from DateAndTime
This issue is related to removing logger and poco dependencies from `Mantid::Kernel::DateAndTime` in preparation for moving the module out of Kernel into a lower level library for use elsewhere.

### PR DESCRIPTION
**Pull request by LamarMoore**
Logging and poco dependencies were removed from `DateAndTime` in preparation for moving this library to the lower level `Mantid::Types` library.

**To test:**

Code review and ensure unit tests pass. Would be good to have input on any edge cases I may have missed regarding file loading and date/time formats.

Fixes #20523. 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
@LamarMoore Could you avoid refactoring tons of Doxygen with every PR? It makes the diff hard to read. Maybe this is a setting issue in your IDE?

Could you undo the formatting changes here?

File: Framework/Kernel/src/DateAndTime.cpp [hunk 1]
```diff
@@ -381,30 +375,9 @@ const DateAndTime &DateAndTime::defaultTime() {
  * @param str :: ISO8601 format string: "yyyy-mm-ddThh:mm:ss[Z+-]tz:tz"
  * @param displayLogs :: flag to indiciate if the logs should be displayed
  */
-void DateAndTime::setFromISO8601(const std::string &str, bool displayLogs) {
+void DateAndTime::setFromISO8601(const std::string &str) {
   // Make a copy
   std::string time = str;
-  // Some ARGUS files have an invalid date with a space instead of zero.
```
Comment (SimonHeybrock) [inline Framework/Kernel/src/DateAndTime.cpp]: Does is make sense to add a unit test, checking that this throws now?

**Comment by LamarMoore** [comment]
@SimonHeybrock It is probably clang on my system which reformatting the doxygen. I will look into it.

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
Looks good, but please undo the reformatting of all the doxygen to keep the diff reasonable and avoid merge conflicts.

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h [hunk 2]
```diff
@@ -114,30 +114,24 @@ class MANTID_KERNEL_DLL DateAndTime {
...
-  static bool stringIsISO8601(const std::string &str);
+
+  /// The difference in seconds between standard unix and gps epochs.
+  static const uint32_t EPOCH_DIFF = 631152000;
+
+  /// The epoch for GPS times.
+  static const boost::posix_time::ptime GPS_EPOCH;
+
+  /// Const of one second time duration
+  static const time_duration oneSecond;
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Can we capitalize this for consistency?

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h [hunk 3]
```diff
@@ -114,30 +114,24 @@ class MANTID_KERNEL_DLL DateAndTime {
...
+
+  static time_t utc_mktime(struct tm *utctime);
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Should this in fact be in the helpers namespace?

File: Framework/Kernel/src/DateAndTime.cpp [hunk 4]
```diff
@@ -1,28 +1,28 @@
...
-#include <boost/lexical_cast.hpp>
-#include <boost/date_time/time.hpp>
 #include <boost/date_time/date.hpp>
+#include <boost/date_time/time.hpp>
+#include <boost/lexical_cast.hpp>
 
-#include <math.h>
 #include <exception>
 #include <limits>
+#include <math.h>
```
Comment (martyngigg) [inline Framework/Kernel/src/DateAndTime.cpp]: --> `<cmath>`

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h [hunk 5]
```diff
@@ -5,10 +5,10 @@
 #ifndef Q_MOC_RUN
 #include <boost/date_time/posix_time/posix_time.hpp>
 #endif
-#include <time.h>
 #include <iosfwd>
 #include <stdint.h>
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Can we convert these to the C++ versions: `<cmath>` & `<cstdint>`?

File: Framework/Kernel/src/DateAndTime.cpp (hunk 1 above)
Comment (martyngigg) [reply Framework/Kernel/src/DateAndTime.cpp]: Agreed.

File: Framework/Kernel/test/DateAndTimeTest.h [hunk 6]
```diff
@@ -8,20 +8,20 @@
...
+#include <cxxtest/TestSuite.h>
+#include <sstream>
 #include <sys/stat.h>
 #include <time.h>
-#include <sstream>
 
 using namespace Mantid;
 using namespace Mantid::Kernel;
 
+using std::cout;
```
Comment (martyngigg) [inline Framework/Kernel/test/DateAndTimeTest.h]: Do we need `cout` in the test?

**Review by martyngigg — CHANGES_REQUESTED** [review]
Generally looks good. Just a few minor comments

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h (hunk 3 above)
Comment (LamarMoore) [reply Framework/Kernel/inc/MantidKernel/DateAndTime.h]: It is used internally (within `DateAndTime`) as well.

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h [hunk 7]
```diff
@@ -5,10 +5,10 @@
 #ifndef Q_MOC_RUN
 #include <boost/date_time/posix_time/posix_time.hpp>
 #endif
-#include <time.h>
+#include <cstdint>
 #include <iosfwd>
-#include <stdint.h>
 #include <string>
+#include <time.h>
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Should this be `<ctime>`?

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h [hunk 8]
```diff
@@ -114,30 +114,24 @@ class MANTID_KERNEL_DLL DateAndTime {
   static void createVector(const DateAndTime start,
                            const std::vector<double> &seconds,
                            std::vector<DateAndTime> &out);
-  static bool stringIsISO8601(const std::string &str);
+
+  /// The difference in seconds between standard unix and gps epochs.
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Is there a reason we have one defined in the header and the others in the `.cpp` file?

File: Framework/Kernel/src/DateAndTime.cpp [hunk 9]
```diff
@@ -379,32 +371,10 @@ const DateAndTime &DateAndTime::defaultTime() {
 /** Sets the date and time using an ISO8601-formatted string
  *
  * @param str :: ISO8601 format string: "yyyy-mm-ddThh:mm:ss[Z+-]tz:tz"
- * @param displayLogs :: flag to indiciate if the logs should be displayed
  */
-void DateAndTime::setFromISO8601(const std::string &str, bool displayLogs) {
+void DateAndTime::setFromISO8601(const std::string &str) {
   // Make a copy
   std::string time = str;
```
Comment (martyngigg) [inline Framework/Kernel/src/DateAndTime.cpp]: We could change the signature of `setFromISO8601` to `setFromISO8601(std::string time)` and remove the manual copy? That way we let the compiler do a move if possible: https://stackoverflow.com/questions/46028709/why-is-passing-by-value-if-a-copy-is-needed-recommended-in-c11-if-a-const-re

File: Framework/Kernel/test/DateAndTimeTest.h [hunk 10]
```diff
@@ -8,20 +8,19 @@
 #ifndef DATEANDTIMETEST_H_
 #define DATEANDTIMETEST_H_
 
-#include <cxxtest/TestSuite.h>
 #include "MantidKernel/DateAndTime.h"
 #include "MantidKernel/System.h"
+#include <cxxtest/TestSuite.h>
+#include <sstream>
 #include <sys/stat.h>
 #include <time.h>
```
Comment (martyngigg) [inline Framework/Kernel/test/DateAndTimeTest.h]: `<ctime>`

File: Framework/Kernel/test/DateAndTimeTest.h [hunk 11]
```diff
@@ -58,6 +57,10 @@ class DateAndTimeTest : public CxxTest::TestSuite {
     TS_ASSERT_EQUALS(d, expected);
   }
 
+  void test_constructor_fails_invalid_string() {
```
Comment (martyngigg) [inline Framework/Kernel/test/DateAndTimeTest.h]: Could we add some tests with strings that look a bit like a date time, i.e the old strings that would have passed but will now fail?

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h (hunk 8 above)
Comment (LamarMoore) [reply Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Yes, we cannot define non-integral types in the header. Builds fail.

File: Framework/Kernel/src/DateAndTime.cpp (hunk 9 above)
Comment (LamarMoore) [reply Framework/Kernel/src/DateAndTime.cpp]: `str` is used within the exception handling (which is why it is preserved)

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h (hunk 8 above)
Comment (LamarMoore) [reply Framework/Kernel/inc/MantidKernel/DateAndTime.h]: @martyngigg I do not think we can initialise non-integral types in a header without some compilers complaining.

File: Framework/Kernel/src/DateAndTime.cpp (hunk 9 above)
Comment (martyngigg) [reply Framework/Kernel/src/DateAndTime.cpp]: Oh yes, I missed that. In that case we should stick with the original `const &` version to avoid having 2 copies when an l-value input is used.

Either that or we remove the `str` from the exception and just throw the `std::invalid_argument` without including the string. The caller could then decide whether to include it in their error message if they want. It seems a shame that most of the time when the string is valid we are paying the price of a copy just for the error message.

File: Framework/Kernel/inc/MantidKernel/DateAndTime.h (hunk 8 above)
Comment (martyngigg) [reply Framework/Kernel/inc/MantidKernel/DateAndTime.h]: Yes, static non-integral types have to be defined out of line. I was suggesting that we be consistent and put `EPOCH_DIFF` in `.cpp` too. I'm not overly concerned though.

**Review by martyngigg — APPROVED** [review]
I'm happy with this to be merged now.

**Comment by SimonHeybrock** [comment]
@LamarMoore The diff is showing unrelated changes from our `LoadEventNexus` feature branch? If you remove those commits I could merge this.

**Comment by SimonHeybrock** [comment]
@LamarMoore Removed unrelated commits as dicussed on slack, looks ok, and github show previous review comments. 👍

**Review by SimonHeybrock — APPROVED** [review]

File: Framework/Kernel/src/DateAndTimeHelpers.cpp [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+bool stringIsISO8601(const std::string &date) {
+  Poco::DateTime dt;
+  int tz_diff;
+  return Poco::DateTimeParser::tryParse(Poco::DateTimeFormat::ISO8601_FORMAT,
+                                        date, dt, tz_diff);
+}
+
+/** Verifies whether or not a string conforms to ISO8601. Corrects the string
+ *if it does not and is of the ARGUS file date/time format
+ *
```
Comment (OwenArnold) [inline Framework/Kernel/src/DateAndTimeHelpers.cpp]: For future readers, i think it would be useful to have something in the comments here, or for `isArgusDateTime` which gives and example of what that date-time string is like. and what the corresponding sanitized string ends up like. I wouldn't object if this was done as a separate PR, as I know that this PR is holding up others.

File: Framework/Kernel/test/DateAndTimeHelpersTest.h [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    TS_ASSERT(
+        DateAndTimeHelpers::stringIsISO8601("1990-01-02T03:04:02.000+05:30"));
+    TS_ASSERT(DateAndTimeHelpers::stringIsISO8601("1990-01-02 03:04"));
+    TS_ASSERT(DateAndTimeHelpers::stringIsISO8601("1990-01-02"));
+    TS_ASSERT(DateAndTimeHelpers::stringIsISO8601("1822-01-02"));
+
+    TS_ASSERT(!DateAndTimeHelpers::stringIsISO8601("January 1, 2345"));
+    TS_ASSERT(!DateAndTimeHelpers::stringIsISO8601("2010-31-56"));
+    TS_ASSERT(!DateAndTimeHelpers::stringIsISO8601("1990-01-02 45:92:22"));
+  }
```
Comment (OwenArnold) [inline Framework/Kernel/test/DateAndTimeHelpersTest.h]: As a follow on from my comment above. is there any reason that the other methods in this namespace can't be tested? I think that would be useful.

File: Framework/Kernel/src/DateAndTimeHelpers.cpp (hunk 12 above)
Comment (LamarMoore) [reply Framework/Kernel/src/DateAndTimeHelpers.cpp]: This will be added in this PR as changes are made there for DateAndTime anyway https://github.com/mantidproject/mantid/pull/20481

File: Framework/Kernel/test/DateAndTimeHelpersTest.h (hunk 13 above)
Comment (LamarMoore) [reply Framework/Kernel/test/DateAndTimeHelpersTest.h]: Same as above. Will be added in this PR https://github.com/mantidproject/mantid/pull/20481

**Comment by martyngigg** [comment]
@OwenArnold I'll merge this once the builds are happy
